compiler/crates/relay-typegen/src/lib.rs

```diff
@@ -643,55 +669,137 @@ impl<'schema, 'config> TypeGenerator<'schema, 'config> {
+                } else {
+                    Some((concrete_type, selections))
+                }
+            })
+            .collect::<IndexMap<_, _>>();
+
+        let mut concrete_types: Vec<_> = Vec::new();
+        let type_fields_present_for_union = !by_concrete_type.is_empty()
             && (base_fields.values().any(TypeSelection::is_typename)
-                || by_concrete_type.values().all(has_typename_selection))
-        {
+                || by_concrete_type.values().all(has_typename_selection));
+
+        if type_fields_present_for_union {
             let mut typename_aliases = IndexSet::new();
-            for (concrete_type, selections) in by_concrete_type {
-                types.push(
+            for (concrete_type, selections) in &by_concrete_type {
+                let selection_names = selections
+                    .iter()
+                    .map(|selection| selection.schema_name)
+                    .collect::<Vec<_>>();
+
+                concrete_types.push(
                     group_refs(
                         base_fields
                             .iter()
-                            .map(|(_, v)| v.clone())
-                            .chain(selections.into_iter())
+                            .map(|(_, v)| v)
+                            .filter(|v| {
+                                v.is_typename() && !selection_names.contains(&v.schema_name)
+                            })
+                            .cloned()
+                            .chain(selections.clone().into_iter())
                             .collect(),
                     )
                     .into_iter()
                     .map(|selection| {
                         if selection.is_typename() {
                             typename_aliases.insert(selection.key);
                         }
-                        self.make_prop(selection, unmasked, Some(concrete_type))
+                        self.make_prop(selection, unmasked, Some(*concrete_type), None)
                     })
                     .collect(),
                 );
             }
 
-            // It might be some other type then the listed concrete types. Ideally, we
-            // would set the type to diff(string, set of listed concrete types), but
-            // this doesn't exist in Flow at the time.
-            types.push(
-                typename_aliases
-                    .iter()
-                    .map(|typename_alias| Prop {
-                        key: *typename_alias,
-                        read_only: true,
-                        optional: false,
-                        value: AST::OtherTypename,
-                    })
-                    .collect(),
-            );
-        } else {
-            let mut selection_map = selections_to_map(hashmap_into_value_vec(base_fields), false);
+            // It might be some other type then the listed concrete types. We try to
+            // figure out which types remain here.
```

Comment:
Is there benefit to this?

This kindof blew up with many pages of types for some huge interfaces. We would need to make this an option, or maybe only do if "no future proof types" is set. There's probably no point in combining the listing out of typenames when also including ther `%other` sentinel value?

```
`'%other' | 'SomeOtherType' | 'YetAnotherType'
```

Reply:
You're right, probably not. I changed to only produce either %other (future proof) or the possible typenames (no future proofness).